### PR TITLE
Preparations for Trivial Support For Resharding (part1)

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Some fixes and refactoring as as a preparation for #45766
To reduce the patch size in future and to reduce probability of conflicts.

_P.S. let me know if you would prefer this to be submitted as a more logical separated patches_

Cc: @alexey-milovidov

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
This is an automated comment for commit 54cd1713bb82205e0e59452d4a5db2503babf1c8 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/52068/54cd1713bb82205e0e59452d4a5db2503babf1c8/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by azat** [comment]
>Stress test (tsan) — Killed by signal (in clickhouse-server.log)

- https://github.com/ClickHouse/ClickHouse/issues/52013#issuecomment-1634086108

**Comment by azat** [comment]
>Stateless tests (release, analyzer) — Some tests restarted, fail: 1, passed: 5390, skipped: 34

New test failed for analyzer, will take a look:

```
2023-07-13 03:37:34 --- /usr/share/clickhouse-test/queries/0_stateless/02813_tree_rewriter_virtual_columns.reference	2023-07-13 03:33:30.550705647 +0800
2023-07-13 03:37:34 +++ /tmp/clickhouse-test/0_stateless/02813_tree_rewriter_virtual_columns.stdout	2023-07-13 03:37:34.186126997 +0800
2023-07-13 03:37:34 @@ -1,2 +1,2 @@
2023-07-13 03:37:34 -"number","_partition_id"
2023-07-13 03:37:34 -0,"all"
2023-07-13 03:37:34 +"_partition_id_0"
2023-07-13 03:37:34 +"all"
```

**Comment by azat** [comment]
>Integration tests (tsan) [4/6] — fail: 1, passed: 274, flaky: 0

#52083

File: src/Planner/PlannerJoinTree.cpp [hunk 1]
```diff
@@ -182,6 +182,9 @@ bool applyTrivialCountIfPossible(
         return false;
 
     const auto & storage = table_node.getStorage();
+    if (!storage->supportsTrivialCountOptimization())
```
Comment (azat) [inline src/Planner/PlannerJoinTree.cpp:185]: This also fixes the `test_materialized_mysql_database` test under analyzer - https://s3.amazonaws.com/clickhouse-test-reports/52096/9d728161f1d8e82bdb7537faa53ca084fbc55643/integration_tests__asan__analyzer__[4_6].html

**Review by alexey-milovidov — APPROVED** [review]

**Comment by azat** [comment]
@alexey-milovidov both times it failed in same place, merge with master will not help

```
 [ 70436 ] {} <Fatal> BaseDaemon: ########################################
 [ 70436 ] {} <Fatal> BaseDaemon: (version 23.6.2.18 (official build), build id: D0E83BD1974B9B4B1FE300F9D23222CFF56067E2, git hash: 89f39a7ccfe0c068c03555d44036042fc1c09d22) (from thread 2318) (query_id: 6820d9ec-c534-49a2-947f-ead82238e795) (query: 
 [ 70436 ] {} <Fatal> BaseDaemon: Address: 0x7efb02911428. Access: read. Address not mapped to object.
 [ 70436 ] {} <Fatal> BaseDaemon: Stack trace: 0x0000000014b9f0f7 0x0000000014ba011d 0x0000000012c72fc3
 [ 70436 ] {} <Fatal> BaseDaemon: 2.1. inlined from ./build_docker/./src/Common/CurrentThread.cpp:50: DB::CurrentThread::getProfileEvents()
 [ 70436 ] {} <Fatal> BaseDaemon: 2.2. inlined from ./build_docker/./src/Common/ProfileEvents.cpp:619: ProfileEvents::increment(StrongTypedef<unsigned long, ProfileEvents::EventTag>, unsigned long)
 [ 70436 ] {} <Fatal> BaseDaemon: 2. ./build_docker/./src/Client/ConnectionEstablisher.cpp:72: DB::ConnectionEstablisher::run(PoolWithFailoverBase<DB::IConnectionPool>::TryResult&, String&) @ 0x0000000014b9f0f7 in /usr/lib/debug/usr/bin/clickhouse.debug
 [ 70436 ] {} <Fatal> BaseDaemon: 3. ./build_docker/./src/Client/ConnectionEstablisher.cpp:129: DB::ConnectionEstablisherAsync::Task::run(std::function<void (int, Poco::Timespan, DB::AsyncEventTimeoutType, String const&, unsigned int)>, std::function<void ()>) @ 0x0000000014ba011d in /usr/lib/debug/
 [ 70436 ] {} <Fatal> BaseDaemon: 4.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:818: ~__policy_func
 [ 70436 ] {} <Fatal> BaseDaemon: 4.2. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1174: ~function
 [ 70436 ] {} <Fatal> BaseDaemon: 4.3. inlined from ./build_docker/./src/Common/AsyncTaskExecutor.cpp:77: DB::AsyncTaskExecutor::Routine::operator()(std::function<void ()>)
 [ 70436 ] {} <Fatal> BaseDaemon: 4.4. inlined from ./build_docker/./src/Common/Fiber.h:70: Fiber::RoutineImpl<DB::AsyncTaskExecutor::Routine>::operator()(boost::context::fiber&&)
 [ 70436 ] {} <Fatal> BaseDaemon: 4.5. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394: decltype(std::declval<Fiber::RoutineImpl<DB::AsyncTaskExecutor::Routine>&>()(std::declval<boost::context::fiber>())) std::__invoke[abi:v15000]<Fiber::RoutineImpl<DB::As
 [ 70436 ] {} <Fatal> BaseDaemon: 4.6. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:531: std::invoke_result<Fiber::RoutineImpl<DB::AsyncTaskExecutor::Routine>&, boost::context::fiber>::type std::invoke<Fiber::RoutineImpl<DB::AsyncTaskExecutor::Routine>&, bo
 [ 70436 ] {} <Fatal> BaseDaemon: 4.7. inlined from ./build_docker/./contrib/boost/boost/context/fiber_fcontext.hpp:140: boost::context::detail::fiber_record<boost::context::fiber, FiberStack&, Fiber::RoutineImpl<DB::AsyncTaskExecutor::Routine>>::run(void*)
 [ 70436 ] {} <Fatal> BaseDaemon: 4. ./build_docker/./contrib/boost/boost/context/fiber_fcontext.hpp:80: void boost::context::detail::fiber_entry<boost::context::detail::fiber_record<boost::context::fiber, FiberStack&, Fiber::RoutineImpl<DB::AsyncTaskExecutor::Routine>>>(boost::context::detail::tran
 [ 70436 ] {} <Fatal> BaseDaemon: Integrity check of the executable successfully passed (checksum: AD6642440A13FE688E9E670BB288E827)
 [ 70436 ] {} <Fatal> BaseDaemon: Report this error to https://github.com/ClickHouse/ClickHouse/issues
 [ 70436 ] {} <Fatal> BaseDaemon: Changed settings: min_compress_block_size = 703890, max_compress_block_size = 1293552, max_block_size = 72876, max_insert_threads = 7, max_threads = 1, max_read_buffer_size = 686176, max_distributed_connections = 1, connect_timeout_with_failover_ms = 2000, connect_t
 [ 1029 ] {} <Fatal> Application: Child process was terminated by signal 11.
```

- https://github.com/ClickHouse/ClickHouse/issues/51313

**Comment by alexey-milovidov** [comment]
There is a conflict which I cannot resolve from the UI - let's merge with master.

**Comment by azat** [comment]
>There is a conflict which I cannot resolve from the UI - let's merge with master.

Done

**Comment by alexey-milovidov** [comment]
What's with the debug stress test?

**Comment by azat** [comment]
>What's with the debug stress test?

Something interesting in jemalloc - https://github.com/ClickHouse/ClickHouse/issues/52487

**Comment by azat** [comment]
Test failures unrelated:

>Integration tests (asan) [5/6] — fail: 1, passed: 440

- `test_merge_tree_azure_blob_storage/test.py::test_table_manipulations` - https://github.com/ClickHouse/ClickHouse/issues/39700

>Upgrade check (tsan) — Logical error thrown (see clickhouse-server.log or logical_errors.txt)

```
/var/log/clickhouse-server/clickhouse-server.err.log:2023.07.24 08:42:05.060947 [ 1675 ] {f79024f9-61f3-40fe-8632-110c6612c30f} <Error> executeQuery: Code: 49. DB::Exception: Broken buckets. its size(5) is not power of 2: While executing FillingRightJoinSide. (LOGICAL_ERROR) (version 23.6.2.18 (offi
```

This has been fixed in upstream already.

**Pull request merged**